### PR TITLE
Add expenses with an equal default split

### LINKED ISSUE #6: Manually Add Expenses
As a group member,

I want to manually record a new expense with the amount and description,

So that the shared balance reflects who paid for what.

### Acceptance Criteria

- [ ] **AC1 — Successful expense creation**
  Given a group member is on the group's "Add expense" form
  When they enter an amount, description, payer and date, and save
  Then the expense is added to the group's expense list
  And each member's balance in the group updates to reflect it

- [ ] **AC2 — Required fields**
  Given a group member is on the "Add expense" form
  When they submit with amount, description or payer missing
  Then the expense is not created
  And an inline error is shown on each missing field

- [ ] **AC3 — Amount validation**
  Given a group member is on the "Add expense" form
  When they enter an amount that is zero, negative, or not a valid number
  Then the expense is not created
  And an error states the amount must be a positive number

- [ ] **AC4 — Default split**
  Given a group member creates an expense without choosing a split method
  When the expense is saved
  Then it is split equally among all current group members

- [ ] **AC5 — Payer must be a group member**
  Given a group member is creating an expense
  When they select a payer
  Then only current members of the group are selectable as payer

- [ ] **AC6 — Date defaults sensibly**
  Given a group member is on the "Add expense" form
  When they save without selecting a date
  Then today's date is used
  And they may override it with a past date, but not a future date

- [ ] **AC7 — Expense appears immediately**
  Given an expense has just been saved
  When any group member opens the group's expense list
  Then the new expense appears with amount, description, payer and date

- [ ] **AC8 — Only members can add expenses**
  Given a user is not a member of the group
  When they attempt to add an expense to it via the API or UI
  Then the request is rejected with an authorisation error

### PR DESCRIPTION
**Pull request by mirosgithub**
## Summary
Adds the write path for expenses. A group member can record an amount, description, payer and date, and the expense is split equally across the group so every member's balance reflects it.

## Associated issue
Closes #6

## Changes

- `POST /groups/{id}/expenses` and `GET /groups/{id}/expenses`, both restricted to members of the group
- Each new expense is split equally across the group's current members and applied to the `net_balance` field #47 added, as agreed in the coordination comment on #4
- Cents left over by an uneven division go to the lowest user ids, so the shares always sum to the amount
- `V4__create_expenses_table.sql`, since #47 holds V3
- An "Add an expense" form at `/groups/:id/expenses/new` with the payer drawn from the group's members, a date defaulting to today that rejects future dates, and an error on each invalid field
- The group page now lists the group's expenses and each member's balance

## Testing
`./mvnw test` in `backend`: 37 tests pass (11 new) in `ExpenseIntegrationTest` covering AC1 through AC8.
`npm test` in `frontend`: 34 tests pass, 8 in `AddExpense.test.jsx` and 2 in the new `expensesApi.test.js`, with `GroupPage.test.jsx` extended for the expense list and the balances.

Checked manually against a running backend and dev server, as user `miro` in a two-member group: an empty form shows an error per field, a zero or negative amount is refused, the payer dropdown holds only the group's members, a future date is refused and a past one accepted, and a 5.50 expense paid by the other member left balances of -2.75 and +2.75 that sum to zero.

AC7's second-member view and AC8's non-member rejection cannot be reached through the UI while `StubCurrentUserProvider` signs every request in as the first user, so both are covered by `ExpenseIntegrationTest` only.

## Dependencies
Depended on #4, merged as #47, and this branch is rebased onto it. Blocks #8, which takes over the split logic for split methods and subsets. No new dependency in `backend/pom.xml` or `frontend/package.json`.

## Reviewer notes
The expense endpoints throw `GroupAccessDeniedException` (403) for a non-member, matching the convention #4 set for the member endpoints rather than the 404 `getGroup` still returns.

`readError` and `requirePositiveInteger` are now exported from `frontend/src/api/groups.js` so `expenses.js` can use them instead of holding a copy.

The split writes straight to `net_balance` with no per-expense share rows. #8 can add them if it needs them.

## Generative AI
Not applicable.

### CODE REVIEW

File: frontend/src/pages/AddExpense.jsx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+function today() {
+    return new Date().toISOString().slice(0, 10);
+}
+
+function validate({ amount, description, paidByUserId, expenseDate }) {
+    const errors = {};
+
+    if (amount.trim() === '') {
+        errors.amount = 'Amount is required';                          // AC2
+    } else if (!(Number(amount) > 0)) {
```
Comment (mirosgithub) [inline frontend/src/pages/AddExpense.jsx:21]: SonarCloud reports an issue here, but I am not sure if this should be changed. `Number("str")` is `NaN`, and `NaN > 0` is `false` so the current operator catches a string input, while changing it to `NaN <= 0` cannot catch it as it also evaluates to `false`. I would appreciate reviewers' input on this.

Comment (Jago-Sutherland) [reply frontend/src/pages/AddExpense.jsx:21]: !(Number(amount) > 0) is logically equivalent to Number(amount) <= 0. Sonar recommends swapping to avoid the unnecessary logical negation. I believe this does not need to be changed, as  ! > 0 is more clear in it's purpose than <= 0.

**Review by Jago-Sutherland — COMMENTED** [review]
I haven't finished reading through the code, but from how and understand it, and the description of the pull request, it seems like you have also implemented the functionality of issue #8 in this feature.

**Comment by mirosgithub** [comment]
> I haven't finished reading through the code, but from how and understand it, and the description of the pull request, it seems like you have also implemented the functionality of issue #8 in this feature.

You are right, thanks for raising this!

#6's AC4 says an expense saved without a split method is split equally across all current members, which implies #8's AC1. #6's AC1 requires each member's balance to update, which requires #8's AC5 and AC4, since the payer would otherwise be charged a share as a debt to themselves. Those three were kind of unavoidable in #6, but I should have noted that in this PR.

I treated the rounding as part of keeping the splits sum to the original amount, but it matches #8's AC2.

#8's AC3, AC6 and AC7 are not covered in this PR so still need to be implemented.

We could tick or remove the covered AC's in #8 before implementing the rest, or undo the relevant changes in this PR. It is your call since you are assigned to #8.

**Comment by Jago-Sutherland** [comment]
I don't think it's worth undoing all of your work just so I can do it instead. I will finish reviewing this, and then tick off the completed acceptance criteria for #8.

**Review by yd35 — CHANGES_REQUESTED** [review]
The main feature looks good and I verified the full build locally. I found two edge cases that need fixing before approval:

- An amount such as `0.004` passes the positive amount validation but gets rounded to `0.00` before it is saved. Please reject amounts that become zero after rounding and add a test for this.
- The form gets the current date using `toISOString()`, which uses UTC. Shortly after midnight in New Zealand, this returns the previous date. Please calculate the date using the local timezone and add a test for it.

The overlap with Issue #8 makes sense because Issue #6 AC1 and AC4 already require the default equal split. I do not think that part needs to be removed.

Minor: the current test totals are 36 backend tests and 30 frontend tests, so the Testing section can be updated.

Once these are fixed and the checks pass again, I am happy to review it again.

**Review by Jago-Sutherland — COMMENTED** [review]
Other than the issues mentioned by @yd35 above, this looks good.

**Review by Jago-Sutherland — APPROVED** [review]
Looks good.

**Review by yd35 — APPROVED** [review]
Re-reviewed the latest fixes. Amounts below 0.01 are now rejected, the form uses the local date, and both edge cases have regression tests. I ran the full backend verification with 37 tests, the frontend tests with 34 tests, and the frontend build. Everything passed, including the GitHub checks. Looks good to me.
